<style>
  .parameter {
    border-top: 4px solid lightblue;
    background-color: rgba(173, 216, 230, 0.2);
    padding: 4px;
    display: inline-block;
    font-weight: bold;
  }
</style>
# Integration methods implemented in panpipes

The panpipes integration pipeline implements a variety of tools to batch correct individual modalities and/or integrate across modalities to produce a reduced dimension representation of the dataset.<br>
There are different tools available for each modality such as RNA (also referred to as GEX), PROT (can be referred to as ADT) and ATAC which can be run as required before running `panpipes integration make merge_batch_correction`
ro create the final object with the reduced dimension represented.<br> 

The ideal way to run `panpipes integration` is to use the output `MuData`file from `panpipes preprocess` since it will already be in the required format. 
However, if using independent MuData the object should contain normalised data in the X slot of each modality, a ‘raw_counts’ layer in each modality, and a sample_id column in each slot of the obs and the outer obs.

| Method    | type of integration         | modalities      | code                                                                       | references                                                                                                                                                                                                                          | benchmarks paper                                                                                                                                                                                                                                                         |
|-----------|-----------------------------|-----------------|----------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| harmony   | unimodal (batch correction) | rna, atac, prot | https://github.com/immunogenomics/harmony                                  | Korsunsky I, Millard N, Fan J, Slowikowski K, Zhang F, Wei K, et al. Fast, sensitive and accurate integration of single-cell data with Harmony. Nat Methods. 2019;16:1289–96.                                                       | Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M.F., Strobl, D.C., Zappia, L., Dugas, M., Colomé-Tatché, M. and Theis, F.J., 2022. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, 19(1), pp.41-50. |
| BBKNN     | unimodal (batch correction) | rna, atac, prot | https://github.com/Teichlab/bbknn                                          | Polański K, Young MD, Miao Z, Meyer KB, Teichmann SA, Park J-E. BBKNN: fast batch alignment of single cell transcriptomes. Bioinformatics. 2020;36:964–5                                                                            | Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M.F., Strobl, D.C., Zappia, L., Dugas, M., Colomé-Tatché, M. and Theis, F.J., 2022. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, 19(1), pp.41-50. |
| Scanorama | unimodal (batch correction) | rna             | https://github.com/brianhie/scanorama                                      | Hie B, Bryson B, Berger B. Efficient integration of heterogeneous single-cell transcriptomes using Scanorama. Nat Biotechnol. 2019;37:685–91                                                                                        | Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M.F., Strobl, D.C., Zappia, L., Dugas, M., Colomé-Tatché, M. and Theis, F.J., 2022. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, 19(1), pp.41-50. |
| scVI      | unimodal (batch correction) | rna             | https://github.com/scverse/scvi-tools                                      | Gayoso A, Lopez R, Xing G, Boyeau P, Pour Amiri VV, Hong J, et al. A Python library for probabilistic analysis of single-cell omics data. Nat Biotechnol. 2022;40:163–6.                                                            | Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Müller, M.F., Strobl, D.C., Zappia, L., Dugas, M., Colomé-Tatché, M. and Theis, F.J., 2022. Benchmarking atlas-level data integration in single-cell genomics. Nature methods, 19(1), pp.41-50. |
| MultiVI   | multimodal                  | atac, rna       | https://github.com/scverse/scvi-tools                                      | Ashuach, T., Gabitto, M.I., Koodli, R.V., Saldi, G.A., Jordan, M.I. and Yosef, N., 2023. MultiVI: deep generative model for the integration of multimodal data. Nature Methods, 20(8), pp.1222-1231.                                | Lee, M.Y., Kaestner, K.H. and Li, M., 2023. Benchmarking algorithms for joint integration of unpaired and paired single-cell RNA-seq and ATAC-seq data. Genome Biology, 24(1), p.244.                                                                                    |
| totalVI   | multimodal                  | prot, rna       | https://github.com/scverse/scvi-tools                                      | Gayoso A, Steier Z, Lopez R, Regier J, Nazor KL, Streets A, et al. Joint probabilistic modeling of single-cell multi-omic data with totalVI. Nat Methods. 2021;18:272–82.                                                           | Makrodimitris, S., Pronk, B., Abdelaal, T. and Reinders, M., 2024. An in-depth comparison of linear and non-linear joint embedding methods for bulk and single-cell multi-omics. Briefings in Bioinformatics, 25(1), p.bbad416.                                          |
| MOFA      | multimodal                  | rna, atac, prot | https://github.com/bioFAM/mofapy2                                          | Argelaguet, R., Arnol, D., Bredikhin, D., Deloro, Y., Velten, B., Marioni, J.C. and Stegle, O., 2020. MOFA+: a statistical framework for comprehensive integration of multi-modal single-cell data. Genome biology, 21(1), pp.1-17. | Lee, M.Y., Kaestner, K.H. and Li, M., 2023. Benchmarking algorithms for joint integration of unpaired and paired single-cell RNA-seq and ATAC-seq data. Genome Biology, 24(1), p.244.                                                                                    |
| WNN       | multimodal                  | rna, atac, prot | https://muon.readthedocs.io/en/latest/api/generated/muon.pp.neighbors.html | Hao Y, Hao S, Andersen-Nissen E, Mauck WM, Zheng S, Butler A, et al. Integrated analysis of multimodal single-cell data. Cell. 2021;184:3573–87.                                                                                    |                                                                                                                                                                                                                                                                          |
